include/exadg/compressible_navier_stokes/postprocessor/hdf5_generator.cpp

```diff
@@ -0,0 +1,118 @@
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *  ______________________________________________________________________
+ */
+
+#include <exadg/compressible_navier_stokes/postprocessor/hdf5_generator.h>
+
+#include <deal.II/numerics/vector_tools.h>
+
+#include <exadg/utilities/print_functions.h>
+
+namespace ExaDG
+{
+namespace CompNS
+{
+template<int dim>
+HDF5Data<dim>::HDF5Data() : write_rho(false), write_rho_u(false), write_rho_E(false)
+{
+}
+
+template<int dim>
+void
+HDF5Data<dim>::print(dealii::ConditionalOStream & pcout) const
+{
+  HDF5DataBase<dim>::print(pcout);
+
+  if(this->write_output && this->evaluation_points.size() > 0)
+  {
+    print_parameter(pcout, "Write rho", write_rho);
+    print_parameter(pcout, "Write rho_u", write_rho_u);
+    print_parameter(pcout, "Write rho_E", write_rho_E);
+  }
+}
+
+template class HDF5Data<2>;
+template class HDF5Data<3>;
+
+template<int dim, typename Number>
+HDF5Generator<dim, Number>::HDF5Generator(MPI_Comm const & comm)
+  : HDF5GeneratorBase<dim, Number>(comm)
+{
+}
+
+template<int dim, typename Number>
+void
+HDF5Generator<dim, Number>::setup(dealii::DoFHandler<dim> const & dof_handler_in,
+                                  dealii::Mapping<dim> const &    mapping_in,
+                                  HDF5Data<dim> const &           hdf5_data_in)
+{
+  this->setup_base(dof_handler_in.get_triangulation(), mapping_in, hdf5_data_in);
+
+  dof_handler = &dof_handler_in;
+  hdf5_data   = hdf5_data_in;
+
+  hyperslab_data.reinit(hdf5_data.evaluation_points.size());
+
+  if(hdf5_data.write_rho)
+    this->add_scalar_quantity_dataset("Rho");
+  if(hdf5_data.write_rho_u)
+    this->add_vector_quantity_datasets("Rho_U");
+  if(hdf5_data.write_rho_E)
+    this->add_scalar_quantity_dataset("Rho_E");
+}
+
+template<int dim, typename Number>
+void
+HDF5Generator<dim, Number>::do_evaluate(VectorType const & solution)
+{
+  if(hdf5_data.write_rho || hdf5_data.write_rho_u || hdf5_data.write_rho_E)
+  {
+    auto const values = dealii::VectorTools::point_values<dim + 2>(this->get_remote_evaluator(),
+                                                                   *dof_handler,
+                                                                   solution);
+    if(hdf5_data.write_rho)
+    {
+      this->extract_component(hyperslab_data, values, 0);
+      this->write_scalar_quantity("Rho", hyperslab_data);
```

Comment:
These two functions seem to appear only in combination. I think it would make sense to combine them into a single function.

Also, I do not understand why `hyperslab_data` is an argument of this function? I think of the following interface:
```
this->write_scalar_quantity(values, 0, "Rho");
```

Reply:
resolved with [7647d7b](https://github.com/exadg/exadg/pull/255/commits/7647d7b3beaddfbed1cba903da224c0bbb5f09a5)